vavr/src/main/java/io/vavr/collection/Iterator.java

```diff
@@ -973,6 +973,38 @@ public T getNext() {
         };
     }
 
+    /**
+     * Creates an iterator that repeatedly invokes the supplier
+     * while it's a {@code Some} and end on the first {@code None}
+     *
+     * @param supplier A Supplier of iterator values
+     * @param <T> value type
+     * @param <A> supplier option type
+     * @return A new {@code Iterator}
+     * @throws NullPointerException if supplier produces null value
+     */
+    static <T, A extends T> Iterator<T> iterate(Supplier<Option<A>> supplier) {
```

Comment:
@nfekete @danieldietrich please review.
Here we should use `A extends T` to make it more convenient to use

Reply:
Looks good to me, but I admit I don't really know the advantage of this version over `static <T> Iterator<T> iterate(Supplier<Option<? extends T> supplier)`. Would you help me out with this?

Reply:
`? extends T` works fine in trivial cases like lambdas, but fails on more complex and frequently occurring cases. For example see `IteratorTest.OptionSupplier`. If it don't implement `Supplier<Option<? extends Integer>>`, a compilation error will be made

Reply:
Just stumbled upon this while trying to achieve inbox-zero :) Sorry for the late reply.
We should take care of uniformly using generic declarations across the project.
Maybe the addition of another parameter is a great trick to circumvent the need of type witnesses in the presence of several `? extends T` declarations - I don't know. It looks a bit like Scala's sort of type declaration now (declaration-site variance vs use-site variance).

I will create an issue to investigate it.

Reply:
I double-checked it. The correct declaration should look like this:

```java
static <T> Iterator<T> iterate(Supplier<? extends Option<? extends T>> supplier) { ... }
```

Then all compile errors disappear. I will do an updating PR.

Thx!